src/gt4py/next/ffront/foast_to_itir.py

```diff
@@ -403,6 +400,28 @@ def _map(self, op, *args, **kwargs):
 
         return im.promote_to_lifted_stencil(im.call(op))(*lowered_args)
 
+    def _process_elements(self, obj, obj_type, process_func, expr, **kwargs):
+        """Recursively applies a processing function to the elements of a structured object, preserving its structure."""
+        if isinstance(obj_type, ts.TupleType):
+            if isinstance(obj, foast.Name) or isinstance(obj, foast.Call):
+                return im.promote_to_lifted_stencil(lambda *elts: im.make_tuple(*elts))(
+                    *[
+                        self._process_elements(
+                            obj, obj_type.types[i], process_func, im.tuple_get(i, expr)
+                        )
+                        for i in range(len(obj.type.types))
+                    ]
+                )
+            elif isinstance(obj, foast.TupleExpr):
+                return im.promote_to_lifted_stencil(lambda *elts: im.make_tuple(*elts))(
+                    *[
+                        self._process_elements(el, el.type, process_func, expr, **kwargs)
+                        for el in obj.elts
+                    ]
+                )
+        else:
+            return self._map(im.lambda_("expr")(process_func(expr)), obj)
```

Comment:
Ok so this was tricky to grasp. Here is a simplified version:
```suggestion
     def _process_elements(self,
                          process_func: Callable[[itir.Expr], itir.Expr],
                          obj: foast.Expr,
                          current_el_type: ts.TypeSpec,
                          current_el_expr: itir.Expr = im.ref("expr"),
        ):
        """Recursively applies a processing function to all primitive constituents of a tuple."""
        if isinstance(current_el_type, ts.TupleType):
            return im.promote_to_lifted_stencil(lambda *elts: im.make_tuple(*elts))(
                *[
                    self._process_elements(
                        process_func, obj, current_el_type.types[i], im.tuple_get(i, current_el_expr)
                    )
                    for i in range(len(obj.type.types))
                ]
            )
        else:
            return self._map(im.lambda_("expr")(process_func(current_el_expr)), obj)
```
What I changed:
 - I renamed `obj_type` to `current_el_type` because this is not (neccesarily) as the name suggests the type of `obj` but the type of the object the function is currently operating on (e.g. a some element).
 - Renamed `expr` to `current_el_expr`. In the original version I was rather confused by the fact that the lambda function in the last line (`self._map(im.lambda_("expr")(process_func(expr)), obj)
`) takes a argument called `expr` and its body consists of what is inside the variable `expr`, but the two have completely different meaning. The former, the symbol `expr`, refers to the value of the entire tuple, while the latter, the ITIR expression stored inside the variable `expr`, refers to the value which `_process_elements` is currently processing. By changing the name it is clear that the two are different things.
- I made `expr` or with the new naming `current_el_expr` have a default value. The caller shouldn't need to know that this needs to be "expr". That is in implementation detail just stemming from a choice of the name of the argument in the lambda function at the end.
- I removed the `isinstance(obj, foast.TupleExpr)` branch of the if. This is not necessary. It generates slightly simpler ITIR code, but I would suggest to remove it for simplicity (in combination with the argument naming it made the function very hard to read). If you want to keep it, use something like this:
```
            elif isinstance(obj, foast.TupleExpr):
                assert current_el_expr == "expr"
                return im.promote_to_lifted_stencil(lambda *elts: im.make_tuple(*elts))(
                    *[
                        self._process_elements(process_func, el, el.type, "expr", **kwargs)
                        for el in obj.elts
                    ]
                )
```
So that it is clear that the current expression is always the top-level "expr" because you also change the `obj` to be an element.
- Shorter docstring. Your's was fine, but too long for the line-length limit we have.

Now... two problems remain:

- Since you are using `self._map` and just pass `obj` which is the entire tuple, you end up with duplicating the lowered `obj` expression many times (ones for every element). If you look at the result of `_process_elements` in the debugger you will see that this expression occurs as many times as you have elements in the tuple. I would suggest to add a todo to fix this later. If you have time to do it great, but we should at least postpone to after your icon4py PR is merged.
- I'm pretty sure this implementation will fail for fields with local dimension. I would suggest to throw a `NotImplemented` exception when this condition is true `type_info.contains_local_field(current_el_type)` in the second last line. This can be addressed together with the previous problem.

Reply:
Thanks for the detailled feedback. The function is simpler now and pass the tests. The TODOs were also added.